h/browser/chrome/lib/hypothesis-chrome-extension.js

```diff
@@ -147,10 +148,20 @@ function HypothesisChromeExtension(dependencies) {
       browserAction.deactivate(tabId);
     }
 
-    settings.then(function(settings) {
-      browserAction.updateBadge(tabId, tab.url, settings.serviceUrl);
+    // Here we're calling uriinfo() for two reasons:
+    // 1. Because we want to call updateBadge() in the then() function.
+    // 2. Because we want to fetch the info for the new URL asap, when we
+    //    call uriinfo() again later (when the user clicks on the browser
+    //    button to activate the sidebar) it will return the already-resolved
+    //    Promise.
+    uriinfo(tab.url).then(function(info) {
+      if (!info.blocked) {
+        browserAction.updateBadge(info.total, tab.id);
+      }
+    }).
+    catch(function() {
+      // Silence console error message about uncaught exception here.
```

Comment:
>  I'm a bit sad that I had to use this, but I think the alternative is
just return {name: 'UriInfoError', ...} and
> then in the catch check the value of .name

Checking the value of .name is actually a reasonable approach. Subclassing
`Error` is indeed awkward in JS. The important thing I think is to return
an object which conforms to the interface of 'Error' interface - namely
having name and message properties.

On 15 October 2015 at 12:45, Sean Hammond <notifications@github.com> wrote:

> In h/browser/chrome/lib/hypothesis-chrome-extension.js
> <https://github.com/hypothesis/h/pull/2610#discussion_r42113763>:
>
> > @@ -147,10 +148,20 @@ function HypothesisChromeExtension(dependencies) {
> >        browserAction.deactivate(tabId);
> >      }
> >
> > -    settings.then(function(settings) {
> > -      browserAction.updateBadge(tabId, tab.url, settings.serviceUrl);
> > +    // Here we're calling uriinfo() for two reasons:
> > +    // 1. Because we want to call updateBadge() in the then() function.
> > +    // 2. Because we want to fetch the info for the new URL asap, when we
> > +    //    call uriinfo() again later (when the user clicks on the browser
> > +    //    button to activate the sidebar) it will return the already-resolved
> > +    //    Promise.
> > +    uriinfo(tab.url).then(function(info) {
> > +      if (!info.blocked) {
> > +        browserAction.updateBadge(info.total, tab.id);
> > +      }
> > +    }).
> > +    catch(function() {
> > +      // Silence console error message about uncaught exception here.
>
> @robertknight <https://github.com/robertknight> How's this for the error
> handling? 6dabcb6
> <https://github.com/hypothesis/h/commit/6dabcb63be7213b2395a03ff51f19fa51c25a220>
>
> I'm a bit sad that I had to use this, but I think the alternative is just return
> {name: 'UriInfoError', ...} and then in the catch check the value of .name
> .
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/hypothesis/h/pull/2610/files#r42113763>.
>